### ISSUE TITLE
Cannot load ntb_netdev, ntb_netdev0 device missing

### ISSUE THREAD

**Issue by ComiXZonE** [sbates130272/switchtec-kernel#58]
Greetings,

I have two hosts that need to communicate with each other, one is a PC and the other one is a TI based CPU. ntb and ntb_hw_switchtec modules are loaded without errors on both sides. However, ntb_netdev0 device seems to be missing on both sides and loading ntb_netdev produces two different errors:
on PC: ntb_netdev: probe of ntb_netdev0 failed with error -5
on TI: modprobe: ERROR: could not insert 'ntb_netdev': no such device

ntb_netdev ends up loading on PC but there is no eth interface created and there is no ntb_netdev0 device present in /dev. On TI, ntb_netdev does not even load and there is no ntb_netdev0 present, neither.

What could be the problem? We have a switchtec switch on a custom board but I have been working with a PM5641 switch eval board before and there were no problems at all and I configured them the same way. Also, chiplink recognizes both hosts in their respective partitions but I'm not sure that the hosts enumerate the ntb endpoint correctly.

Thank you in advance.

**Comment by lsgunth** [comment]
Are you loading the ntb_transport module before ntb_netdev?

Also, there will not be any device in /dev, if ntb_netdev succeeds you'll see a new networking device (ie ethX).

Logan

**Comment by ComiXZonE** [comment]
Well when I was working with PM5641 I used to load only ntb_netdev and it would automatically load ntb_transport as its dependency. The ntb_transport is also loaded in this case.

I know that an eth interface should show up but I thought there should be a device in /dev also popping up.

**Comment by lsgunth** [comment]
Any messages when you load ntb_transport? Can you load ntb_transport and see what's in /sys/kernel/debug/ntb_transport? 

I suspect the configuration of your switch does not have enough resources to create the queue pairs.

Logan

**Comment by ComiXZonE** [comment]
Yes, there is a dmesg message and it's the same one on both sides. It states "Software Queue-Pair Transport over NTB, version 4". There is nothing in /sys/kernel/debug/ntb_transport though.

**Comment by ComiXZonE** [comment]
I have also tried connecting a different PC to the same ARM and the same switch and I get different behavior on both sides. This time the messages are:
on PC: switchtec switchtec0: failed to register ntb device: -12 (so there is no point in trying to even load ntb_netdev)
on ARM: switchtec switchtec0: NTB device registered
               ntb_netdev: probe of ntb_netdev0 failed with error -5

It is kind of random behavior. As you can see now the ARM outputs the same message as the first PC did.

**Comment by ComiXZonE** [comment]
Update: I still sometimes  get a "modprobe: ERROR: could not insert 'ntb_netdev': No such device" on ARM but I cannot pinpoint the exact conditions under which this message is being output or the other one. As I said, pretty random behavior.

**Comment by lsgunth** [comment]
Well -12 is ENOMEM, so it sounds like it's unable to allocate memory. It may possible be unable to allocate DMA coherent memory.  But at that stage in the load it shouldn't need to allocate more than 64KB of coherent memory. However, if there's some kind of memory pressure on your system it explains why it is random.

For the ntb_transport problem, it's erroring out and not creating any queue-pairs and I don't know what's missing. You will probably need to instrument ntb_transport_probe() to find out.

Also, it would be helpful to know what kernel version and if you are using the out-of-tree module what version of that.

**Comment by ComiXZonE** [comment]
I will try to investigate further. The PC runs kernel 4.16.7 with all ntb and switchtec drivers already present in there. ARM runs kernel 4.4.84 with ntb drivers present in there and the switchtec drivers taken from this repo's 'backport_4.4_to_4.7' branch.

**Comment by lgacnik** [comment]
@ComiXZonE Was this ever resolved? I have the exact same problem you described.
